google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java

```diff
@@ -822,11 +832,49 @@ private void put(SpannerException e) {
     }
 
     private PooledSession take() throws SpannerException {
-      SessionOrError s = Uninterruptibles.takeUninterruptibly(waiter);
-      if (s.e != null) {
-        throw newSpannerException(s.e);
+      long currentTimeout = options.getInitialWaitForSessionTimeoutMillis();
+      while (true) {
+        Tracer tracer = Tracing.getTracer();
+        Scope waitScope = tracer.spanBuilder(WAIT_FOR_SESSION).startScopedSpan();
+        try {
+          SessionOrError s = pollUninterruptiblyWithTimeout(currentTimeout);
+          if (s == null) {
+            // End the span with status DEADLINE_EXCEEDED and retry.
+            tracer
+                .getCurrentSpan()
+                .end(EndSpanOptions.builder().setStatus(Status.DEADLINE_EXCEEDED).build());
+            currentTimeout = Math.min(currentTimeout * 2, MAX_SESSION_WAIT_TIMEOUT);
+          } else {
+            if (s.e != null) {
+              throw newSpannerException(s.e);
+            }
+            tracer.getCurrentSpan().end();
```

Comment:
Same here. You don't need to call this explicitly.

Reply:
Changed